src/PluginsSystem/Tools/Microsoft.Performance.Toolkit.Plugins.Cli/Program.cs

```diff
@@ -0,0 +1,226 @@
+            ILogger logger = serviceProvider.GetRequiredService<ILogger<Program>>();
+
+            bool outputSpecified = metadataGenOptions.OutputDirectory != null;
+            string? outputDirectory = outputSpecified ? metadataGenOptions.OutputDirectoryFullPath : Environment.CurrentDirectory;
+            
+            string destMetadataFileName = Path.Combine(outputDirectory!, $"{metadata.Identity}-{Constants.MetadataFileName}");
+            string validDestMetadataFileName = (outputSpecified && metadataGenOptions.Overwrite) ?
+                $"{destMetadataFileName}{Constants.MetadataFileExtension}" : GetValidDestFileName(destMetadataFileName, Constants.MetadataFileExtension);
+
+            string destContentsMetadataFileName = Path.Combine(outputDirectory!, $"{metadata.Identity}-{Constants.ContentsMetadataFileName}");
+            string validDestContentsMetadataFileName = (outputSpecified && metadataGenOptions.Overwrite) ?
+                $"{destContentsMetadataFileName}{Constants.MetadataFileExtension}" : GetValidDestFileName(destContentsMetadataFileName, Constants.MetadataFileExtension);
+
+            try
+            {
+                using (FileStream fileStream = File.Open(validDestMetadataFileName, FileMode.Create, FileAccess.Write, FileShare.None))
+                {
+                    metadataSerializer.Serialize(fileStream, metadata);
+                }
+            }
+            catch (IOException ex)
+            {
+                logger.LogDebug(ex, $"IO exception when writing to {validDestMetadataFileName}.");
+                throw new ConsoleRuntimeException($"Failed to create plugin metadata at {validDestMetadataFileName}.", ex);
+            }
+
+            logger.LogInformation($"Successfully created plugin metadata at {validDestMetadataFileName}.");
+            
+            try
+            {
+                using (FileStream fileStream = File.Open(validDestContentsMetadataFileName, FileMode.Create, FileAccess.Write, FileShare.None))
+                {
+                    contentsMetadataSerializer.Serialize(fileStream, processedDir.ContentsMetadata);
+                }
+            }
+            catch (IOException ex)
+            {
+                logger.LogDebug(ex, $"IO exception when writing to {validDestContentsMetadataFileName}.");
+                throw new ConsoleRuntimeException($"Failed to create plugin contents metadata at {validDestContentsMetadataFileName}.", ex);
+            }
+
+            logger.LogInformation($"Successfully created plugin contents metadata at {validDestContentsMetadataFileName}.");
+            
+            return 0;
+        }
+
+        private static void PackageGenCommon(PackageGenCommonOptions options, ServiceProvider serviceProvider, out ProcessedPluginDirectory processedDir, out PluginMetadata metadata)
+        {
+            ISourceFilesProcessor processor = serviceProvider.GetRequiredService<ISourceFilesProcessor>();
```

Comment:
This is called the "Service Locator" pattern, which in modern times is considered a code smell 😄. This pattern undermines the point of using dependency injection.

Instead, you should be creating classes for each of the different actions that can be performed. Each class should have a set of dependencies in its constructor that get automatically satisifed via dependency injection.